Features/add support for egm

Ciszq opened the pull request.

Ciszq committed:
Created IGrid with GridConstants

Ciszq committed:
Created GridInMemory class, small renames and improvements

Ciszq committed:
Created GridFromFileClass,

+null check in both classes
+new field in constants

Ciszq committed:
Added inheritance for Grids

Ciszq committed:
Renames for better understanding functionality

Ciszq committed:
Chanbges in GridInMemory

now gets converted data

Ciszq committed:
Minor changes in Grid Constants

Added parameter extractor, code refactor, code cleanup

Ciszq committed:
Raw version of Grid Factory

Ciszq committed:
Temporary changes in GridFromFile

Ciszq committed:
Raw version of handling zip files

Ciszq committed:
Removed StreamReaders

Ciszq committed:
Loading data from zip and raw files.

Ciszq committed:
Renames, small improvements

Ciszq committed:
Small renames in PgmGridFactory

Ciszq committed:
Changes /issuecomment/135467341

Ciszq committed:
Removed unused packages, code refactor

Ciszq committed:
Renames

Ciszq committed:
Reduce memory usage

When working with 1' data, receives OutOfMemoryException

Ciszq committed:
Restored changes associated with the pgmdata type

Ciszq committed:
Resolved warnings

Ciszq committed:
Resolved last warning and removed abstract class.

Ciszq committed:
Removed unnecessary IEquatable from PgmDataDescription

Ciszq committed:
Code refactor

Ciszq committed:
PgmDataDescriptionExtractor uses one regex to get all the information

Ciszq committed:
Moved common code to separate class, removed datadescription from interface

Ciszq committed:
Renames in grid factory

Ciszq committed:
Created PgmPreamblesHelper with sample preambles

Ciszq committed:
Added DataDescriptionExtractorTests

regex testing

Ciszq committed:
Changes in DataExtractor

Tests fails on wrong composed regex

Ciszq committed:
Added InternalsVisibleTo and changed description in assemblyInfo

Ciszq committed:
Class and folders renames

Ciszq committed:
Splitted huge regex into smaller ones

Ciszq committed:
Small renames

Ciszq committed:
Added AssertDeep

Ciszq committed:
Changes in AssemblyInfo

Moved some parts to global assembly

Ciszq committed:
extractPreambleFromStream uses custom ReadLine function

Ciszq committed:
Added AssertDeep for PgmDataDescriptionExtractorTests

Ciszq committed:
Merge remote-tracking branch 'rainlabs-eu/master' into feature/add-support-for-egm

Ciszq committed:
Moved Pgm with tests to src folder after merge

Ciszq committed:
Merge remote-tracking branch 'rainlabs-eu/master' into feature/add-support-for-egm

Ciszq committed:
Changed the way of extracting data from the preamble

Concise and easy to understand regex formulas

Ciszq committed:
Fix failing PgmDataDescriptionExtractorTests

Ciszq committed:
Repaired TeamCity compilation error

Reinstalled all NuGets in solution.

kwesolowski commented on src/NSrtm.Core.xTests/NSrtm.Core.xTests.csproj:
```diff
@@ -0,0 +1,119 @@
+    </Reference>
+    <Reference Include="System" />
+    <Reference Include="xunit.abstractions, Version=2.0.0.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
+      <HintPath>\\rldc01.rainlabs.local\UserHomes$\michal.ciszewski\Documents\Visual Studio 2015\Projects\NSrtmPrivateTests\packages\xunit.abstractions.2.0.0\lib\net35\xunit.abstractions.dll</HintPath>
```
Still sth wrong here....

kwesolowski commented on src/NSrtm.Core.xTests/NSrtm.Core.xTests.csproj:
```diff
@@ -0,0 +1,119 @@
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="xunit.core, Version=2.0.0.2929, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
+      <HintPath>\\rldc01.rainlabs.local\UserHomes$\michal.ciszewski\Documents\Visual Studio 2015\Projects\NSrtmPrivateTests\packages\xunit.extensibility.core.2.0.0\lib\portable-net45+win+wpa81+wp80+monotouch+monoandroid+Xamarin.iOS\xunit.core.dll</HintPath>
```
Definitely.

kwesolowski commented on src/NSrtm.Core.xTests/NSrtm.Core.xTests.csproj:
```diff
@@ -0,0 +1,119 @@
+    <PropertyGroup>
+      <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
+    </PropertyGroup>
+    <Error Condition="!Exists('\\rldc01.rainlabs.local\UserHomes$\michal.ciszewski\Documents\Visual Studio 2015\Projects\NSrtmPrivateTests\packages\xunit.core.2.0.0\build\portable-net45+win+wpa81+wp80+monotouch+monoandroid+Xamarin.iOS\xunit.core.props')" Text="$([System.String]::Format('$(ErrorText)', '\\rldc01.rainlabs.local\UserHomes$\michal.ciszewski\Documents\Visual Studio 2015\Projects\NSrtm ...
```
one more.

kwesolowski commented on src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescription.cs:
```diff
@@ -0,0 +1,52 @@
+            _maxValue = maxValue;
+            _preambleLength = preambleLength;
+            _numberOfPoints = _gridGraphHeightPoints * _gridGraphWidthPoints;
+            _longitudeIncrement = _gridGraphWidthPoints / 360.0;
```
So increment is number of values per one degree?

Imho increment is value satisfying:
lat[i] + increment = lat[i+1]..

Ciszq replied:
Done

kwesolowski replied:
Please, just push commit resolving.

kwesolowski commented on src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescriptionExtractor.cs:
```diff
@@ -0,0 +1,87 @@
+        public static readonly IReadOnlyDictionary<string, Regex> Extractor =
+            new Dictionary<string, Regex>
+            {
+                {"MagicNumber", new Regex(@"^(P5)[^$]", RegexOptions.Multiline)},
```
This magic number should be checked as `preamble.Substring(2) = "P5"`, not by looking for all line starts (I think it wuold match file where second line is "P5".

Ciszq replied:
With option SingleLine he treates preamble like a one line. I changed to multiline by mistake.

kwesolowski commented on src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescriptionExtractor.cs:
```diff
@@ -0,0 +1,87 @@
+                {"MagicNumber", new Regex(@"^(P5)[^$]", RegexOptions.Multiline)},
+                {"Offset", new Regex(@"^#[\s]Offset[\s]+(?<offset>.+)$", RegexOptions.Multiline)},
+                {"Scale", new Regex(@"^#[\s]Scale[\s]+(?<scale>.+)$", RegexOptions.Multiline)},
+                {"Origin", new Regex(@"^#[\s]Origin[\s](?<lat>\d+)\D+(?<lon>\d+)", RegexOptions.Multiline)},
```
It does not match direction (for example only values):
lat: 90
lon: 0

So sth is seriously wrong with your unit tests - they occupy more space than actual code, and still such bug is not catched.

Ciszq replied:
Are you sure?

This preamble : P5

# Geoid file in PGM format for the GeographicLib::Geoid class

# Description WGS84 EGM84, 30-minute grid

# URL http://earth-info.nga.mil/GandG/wgs84/gravitymod/wgs84_180/wgs84_180.html

# DateTime 2009-08-29 18:45:02

# RMSBilinearError 0.070

# Offset -108

# MaxBilinearError 1.546

# MaxCubicError 0.274

# RMSCubicError 0.014

# Scale 0.003

# Origin 90 0

# AREA_OR_POINT Point

# Vertical_Datum WGS84

720  361
65535

passed regex tester http://regexstorm.net/tester and my UT. Can you be more specific and give me your preamble version?

kwesolowski commented on src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescriptionExtractor.cs:
```diff
@@ -0,0 +1,87 @@
+                {
+                    lastLine = reader.ReadLine();
+                    preamble.Append(lastLine + "\n");
+                } while (lastLine.StartsWith("#", StringComparison.CurrentCulture));
```
Why you explicitely provide CurrentCulture here? why not Chinese culture?

Ciszq replied:
As explained in Best Practices for Using Strings in the .NET Framework, we recommend that you avoid calling string comparison methods that substitute default values and instead call methods that require parameters to be explicitly specified. To determine whether a string begins with a particular substring by using the string comparison rules of the current culture, call the StartsWith(String, StringComparison) method overload with a value of StringComparison.CurrentCulture for its comparisonType parameter.

From https://msdn.microsoft.com/en-us/library/baketfxw(v=vs.110).aspx

kwesolowski commented on src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescriptionExtractor.cs:
```diff
@@ -0,0 +1,87 @@
+
+        private static string extractPreambleFromStream(Stream stream)
+        {
+            using (var reader = new BinaryReader(stream, Encoding.UTF8, true))
```
Why UTF8?

Ciszq replied:
Because i want to use BinaryReader(Stream, Encoding, Boolean) constructor and this encoding is default encoding for BinaryReader(Stream)

Ciszq committed:
Repaired TeamCity compilation error

NuGet reinstalling failed. Problem connected with settings
in package manager Console ( -ignoreDependencies )

kwesolowski commented on src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescriptionExtractor.cs:
```diff
@@ -0,0 +1,87 @@
+                throw new ArgumentException(String.Format("Can not extract value {0} from the preamble", match.Key));
+            }
+
+            var offset = Convert.ToDouble(matches["Offset"].Groups["offset"].Value, CultureInfo.InvariantCulture);
```
Please use typename.Parse (`double.Parse`). 

For reference:
http://stackoverflow.com/questions/586436/double-tryparse-or-convert-todouble-which-is-faster-and-safer

kwesolowski commented on src/NSrtm.Core/Pgm/PgmGeoidUndulationGridFactory.cs:
```diff
@@ -0,0 +1,67 @@
+
+        public static IPgmGeoidUndulationGrid CreateGeoidUndulationGridInMemory(string filePath)
+        {
+            var zipDirectory = Path.GetDirectoryName(Path.GetDirectoryName(filePath));
```
You were supposed to fix that (do not use not existing paths).

Ciszq replied:
Done

kwesolowski commented on src/NSrtm.Core/Pgm/PgmGeoidUndulationGridFactory.cs:
```diff
@@ -0,0 +1,67 @@
+            return File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+        }
+
+        private static UInt16[] getScaledUndulationFromStream(Stream stream, PgmDataDescription dataDescription)
```
For functions reading from stream use name `Read`, for functions getting from class use `get` if property is bad idea.

Ciszq replied:
Thanks for useful advice.

kwesolowski commented on src/NSrtm.Core/packages.config:
```diff
@@ -1,4 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="JetBrains.Annotations" version="9.1.1" targetFramework="net45" developmentDependency="true"/>
+  <package id="JetBrains.Annotations" version="9.1.1" targetFramework="net45" developmentDependency="true" />
```
If you do not understand sth plase ask, do not change sth like that (it casues Nuget dependency between NSrtm and JetBrains Annotations).

kwesolowski commented on src/NSrtm.Demo/NSrtm.Demo.csproj:
```diff
@@ -53,36 +53,8 @@
       <HintPath>..\packages\MahApps.Metro.1.1.2.0\lib\net45\MahApps.Metro.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="ReactiveUI, Version=6.5.0.0, Culture=neutral, processorArchitecture=MSIL">
```
IMHO this references are required and TC is poiting that out.

Ciszq committed:
Change the way of opening zip files

Use zip not pgm path for compressed data

Ciszq committed:
Use Parse instead ConvertTo, remove unuseful constructor arguments

Ciszq committed:
Fix bug connected with lat/lon step calculation

Renames

Ciszq committed:
Remove local path from Core.xTests

kwesolowski commented on src/NSrtm.Core.xTests/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Please use `GlobalAssemblyInfo`.

kwesolowski commented on src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescriptionExtractor.cs:
```diff
@@ -0,0 +1,87 @@
+
+            var offset = Double.Parse(matches["Offset"].Groups["offset"].Value, CultureInfo.InvariantCulture);
+            var scale = Double.Parse(matches["Scale"].Groups["scale"].Value, CultureInfo.InvariantCulture);
+            var lat = Int32.Parse(matches["Origin"].Groups["lat"].Value);
```
Please match S/N and W/E as separate group, and for S/W use negative values (80S => -80).

kwesolowski commented on src/NSrtm.Core/Pgm/PgmGeoidUndulationGridFactory.cs:
```diff
@@ -0,0 +1,66 @@
+                    }
+                }
+            }
+            else
```
Please check extension here, assuming that "non zip is good" is not best idea

kwesolowski commented on src/NSrtm.Core/packages.config:
```diff
@@ -1,4 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="JetBrains.Annotations" version="9.1.1" targetFramework="net45" developmentDependency="true"/>
+  <package id="JetBrains.Annotations" version="9.1.1" targetFramework="net45" />
```
Please restore `developmentDependency="true"`, and please review PR yourself for changes you do not understand.

Ciszq committed:
Check for wrong origin in preamble

Origin should be (90N 0E)

Ciszq committed:
Remove CA warning from DataExtractor

Ciszq committed:
Namespaces should correspond with the class location

Ciszq committed:
Change Core.xTests langVersion to 4.0

Ciszq committed:
Change targeting framework from 4.5.2 to 4.5.0

Due to Team City fails

Ciszq committed:
Add global Assembly for common data

Ciszq committed:
Improve extension checking algorithm

Wrong filw extension throws InvalidFileTypeException

Ciszq committed:
Code refactor, change package dependency

Ciszq committed:
Refactor AssemblyInfo for Core.xTests

Ciszq committed:
SuppressMessage connected with pgm format

Ciszq committed:
Move CLSCompilant attribute from global assembly to local

Ciszq committed:
Change exception accessibility level to more restrictive

Ciszq committed:
Remove CLSCompliant Attribute from GlobalAssemblyInfo template

Ciszq committed:
Add direction recognition to lat/lon angle

For lat - poles recognition,  lon - side from Prime Meridian

Ciszq committed:
Remove unnecessary checking end of line when parsing pgm magic number.

Ciszq committed:
Remove wrong local path in .csproj

kwesolowski commented on src/NSrtm.Core.xTests/NSrtm.Core.xTests.csproj:
```diff
@@ -0,0 +1,127 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\xunit.core.2.0.0\build\portable-net45+win+wpa81+wp80+monotouch+monoandroid+Xamarin.iOS\xunit.core.props" Condition="Exists('..\packages\xunit.core.2.0.0\build\portable-net45+win+wpa81+wp80+monotouch+monoandroid+Xamarin.iOS\xunit.core.props')" />
+  <Import Project="\\rldc01.rainlabs.local\UserHomes$\michal.ciszewski\Documents\Visual Studio 2015\Projects\NSrtmPrivateTests\packages\xunit.core.2.0.0\build\portable-net45+win+wpa81+wp80+monotouch+monoandroid+Xamarin.iOS\xunit.core.props" Condition="Exists('\\rldc01.rainlabs.local\UserHomes$\michal.ciszewski\Documents\Visual Studio 2015\Projects\NSrtmPrivateTests\packages\xunit.core.2.0.0\build ...
```
Please use Ctrl+F and fix all paths with such mistakes...
